driver/src/main/java/org/neo4j/driver/internal/messaging/request/TransactionMetadataBuilder.java

```diff
@@ -87,7 +78,14 @@ public static Map<String, Value> buildMetadata(
             result.put(BOOKMARKS_METADATA_KEY, value(bookmarks.stream().map(Bookmark::value)));
         }
         if (txTimeoutPresent) {
-            result.put(TX_TIMEOUT_METADATA_KEY, value(txTimeout.toMillis()));
+            var millis = txTimeout.toMillis();
+            if (txTimeout.toNanosPart() % 1_000_000 > 0) {
```

Comment:
Another option for this check is to muliply the millis and check against nanos, mod 1,000,000 is probably clearer though.
```suggestion
            if (txTimeout.toNanosPart() != millis * 1_000_000) {
```
This validation could be in some method `checkForSubmillisecondPrecision` too, just to document for yourself in the future.